{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from python_functions import read_tree_json,get_chr_table,flat_list,remove_duplicates_preserve_order,get_unique_lists,get_tree_edges,clean_output,get_driver_mutations,find_parent\n",
    "import seaborn as sns\n",
    "import plotly.graph_objects as go\n",
    "import networkx as nx\n",
    "from plotly.subplots import make_subplots\n",
    "import warnings\n",
    "import os,sys\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# set correct wd:\n",
    "cwd = os.getcwd()\n",
    "if not cwd.endswith('ANALYSIS'):\n",
    "    os.chdir('./bin/ANALYSIS')\n",
    "    sys.path.append(os.path.abspath('.'))\n",
    "output_directory = '../../'\n",
    "os.makedirs(f'{output_directory}/figures', exist_ok=True)\n",
    "\n",
    "def plot_heatmap_with_tree(tree, alpaca_output,cp_table,clone_info,wgd_calls,chr_table_file,mut_table,allele='A', max_cpn_cap=8):\n",
    "    tumour_id = alpaca_output.tumour_id.iloc[0]\n",
    "    # get colours\n",
    "    color_map = sns.color_palette(\"rocket\", 7, as_cmap=False)\n",
    "    color_map.reverse()\n",
    "    # get seeding clones\n",
    "    if clone_info is not '':\n",
    "        seeding_clones = [f'clone{c}' for c in clone_info[(clone_info['SampleID']==tumour_id) & (clone_info['seedingClones'])].PyCloneCluster_SC]\n",
    "    else:\n",
    "        seeding_clones = []\n",
    "    # get mrca\n",
    "    mrca = tree[0][0]\n",
    "    # get wgd clones\n",
    "    if wgd_calls is not None:\n",
    "        gd_clones = list(wgd_calls[(wgd_calls['tumour_id']==tumour_id) & (wgd_calls.GD==1)].clones)\n",
    "    else:\n",
    "        gd_clones = []\n",
    "    \n",
    "    # drop diploid:\n",
    "    alpaca_output = alpaca_output[alpaca_output.clone != 'diploid']\n",
    "    # add coords:\n",
    "    alpaca_output['chr'] = 'chr' + alpaca_output.segment.str.split('_', expand=True)[0]\n",
    "    alpaca_output['Start'] = alpaca_output.segment.str.split('_', expand=True)[1].astype(int)\n",
    "    alpaca_output['End'] = alpaca_output.segment.str.split('_', expand=True)[2].astype(int)\n",
    "    # modify segment positions to absolute:\n",
    "    alpaca_output = alpaca_output.merge(get_chr_table(chr_table_file))\n",
    "    alpaca_output['abs_start'] = alpaca_output['Start'] + alpaca_output['shift']\n",
    "    alpaca_output['abs_end'] = alpaca_output['End'] + alpaca_output['shift']\n",
    "    alpaca_output = alpaca_output.sort_values(['abs_start'], ascending=False)\n",
    "    alpaca_output.loc[alpaca_output['pred_CN_A'] > max_cpn_cap, 'pred_CN_A'] = max_cpn_cap\n",
    "    alpaca_output.loc[alpaca_output['pred_CN_B'] > max_cpn_cap, 'pred_CN_B'] = max_cpn_cap\n",
    "    \n",
    "    number_of_clones = len(alpaca_output.clone.unique())\n",
    "    max_levels = max([len(b) for b in tree])\n",
    "    tree_with_levels = [dict(zip(b, range(0, len(b)))) for b in tree]\n",
    "    tree_with_levels =   pd.concat([pd.DataFrame(tree_with_levels[x], index=[0]).transpose() for x, _ in enumerate(tree_with_levels)]).reset_index().drop_duplicates().rename(\n",
    "        columns={'index': 'clone', 0: 'level'})\n",
    "    \n",
    "    # make empty y_loc df\n",
    "    clone_y_location = dict(zip(alpaca_output.clone.unique(), range(0, number_of_clones)))\n",
    "    clone_y_location = pd.DataFrame(clone_y_location, index=[0]).transpose().reset_index().rename(columns={'index': 'clone', 0: 'y_loc'})\n",
    "    clone_y_location['y_loc'] = 100\n",
    "    # find sections:\n",
    "    # section is a part of a path that requires its own horizontal space on the final graph\n",
    "    ori_tree = tree.copy()\n",
    "    sections = [[tree[0][0]]]\n",
    "    while len(flat_list(ori_tree)) > 1:\n",
    "        for i, branch in enumerate(ori_tree):\n",
    "            if (branch != []) and (remove_duplicates_preserve_order(flat_list(sections)) != remove_duplicates_preserve_order(flat_list(tree))):\n",
    "                branching_clones = list(pd.Series(flat_list(ori_tree)).value_counts()[pd.Series(flat_list(ori_tree)).value_counts() > 1].index)\n",
    "                if branching_clones == []:\n",
    "                    branching_clones = [mrca]\n",
    "                section_start = max([branch.index(x) for x in branching_clones if x in branch]) + 1\n",
    "                section = branch[section_start:]\n",
    "                ori_tree[ori_tree.index(branch)] = branch[:section_start]\n",
    "                ori_tree = get_unique_lists(ori_tree)\n",
    "                if section != []:\n",
    "                    sections.append(section)\n",
    "    # order sections according to proximity\n",
    "    # start with on of the longest paths\n",
    "    section_termini = [x[-1] for x in sections]\n",
    "    initial_node = [s for s in sections if len(s) == max([len(x) for x in sections])][0][-1]\n",
    "    # simplify tree graph:\n",
    "    simple_tree = [[clone for clone in branch if clone in section_termini] for branch in tree]\n",
    "    \n",
    "    edges = get_tree_edges(simple_tree)\n",
    "    \n",
    "    G = nx.Graph()\n",
    "    for edge in edges:\n",
    "        G.add_edge(edge[0], edge[1], weight=0)\n",
    "    \n",
    "    nodes = [initial_node] + [x for x in list(G.nodes) if x is not initial_node]\n",
    "    distance_to_nodes = dict(nx.all_pairs_shortest_path_length(G))\n",
    "    processed_nodes = [[s for s in sections if s[-1] == initial_node][0][-1]]\n",
    "    \n",
    "    while len(nodes) > 0:\n",
    "        node = processed_nodes[-1]\n",
    "        neighbours = pd.DataFrame(distance_to_nodes[node], index=['val']).transpose()\n",
    "        neighbours.drop(inplace=True, index=processed_nodes)\n",
    "        try:\n",
    "            neighbours.drop(inplace=True, index=node)\n",
    "        except KeyError:\n",
    "            pass\n",
    "        neighbours = neighbours.reset_index().rename(columns={'index': 'clone'}).merge(tree_with_levels, how='left', on='clone')\n",
    "        neighbours['level'] = neighbours['level'].astype(int)\n",
    "        if len(neighbours) > 0:\n",
    "            # if there are to neighbours equaly close, choose the one which has higher level (i.e. is deeper in the tree)\n",
    "            # to do that, multiply proximity by negaive level:\n",
    "            closest_candidates = neighbours[neighbours.val == min(neighbours.val)]\n",
    "            closest_neighbours = closest_candidates[closest_candidates.level == max(closest_candidates.level)]\n",
    "            closest_neighbour = closest_neighbours.clone.values[0]\n",
    "            if closest_neighbour not in processed_nodes:\n",
    "                processed_nodes.append(closest_neighbour)\n",
    "        nodes = [n for n in nodes if n != node]\n",
    "    \n",
    "    sorted_sections_raw = []\n",
    "    for node in processed_nodes:\n",
    "        sorted_sections_raw.append([s for s in sections if s[-1] == node][0])\n",
    "    \n",
    "    # join single-clone sections below MRCA with their descendants:\n",
    "    sorted_sections = []\n",
    "    below_MRCA = True\n",
    "    skip_element = False\n",
    "    for ss in sorted_sections_raw:\n",
    "        if skip_element:\n",
    "            skip_element = False\n",
    "            continue\n",
    "        if mrca in ss:\n",
    "            below_MRCA = False\n",
    "        if below_MRCA:\n",
    "            if len(ss) == 1:\n",
    "                if sorted_sections_raw.index(ss) < len(sorted_sections_raw) - 1:\n",
    "                    joined = ss + sorted_sections_raw[sorted_sections_raw.index(ss) + 1]\n",
    "                    sorted_sections.append(joined)\n",
    "                else:\n",
    "                    sorted_sections[-1] = sorted_sections[-1] + ss\n",
    "                skip_element = True\n",
    "            else:\n",
    "                sorted_sections.append(ss)\n",
    "        else:\n",
    "            sorted_sections.append(ss)\n",
    "    \n",
    "    # assign y location on the plot to sorted sections:\n",
    "    available_y_locs = range(0, number_of_clones)\n",
    "    below_MRCA = True\n",
    "    for section in sorted_sections:\n",
    "        if mrca in section:\n",
    "            below_MRCA = False\n",
    "        locs_for_this_section = list(available_y_locs[:len(section)])\n",
    "        if below_MRCA:\n",
    "            locs_for_this_section = list(reversed(locs_for_this_section))\n",
    "        locs_for_this_section_dict = dict(zip(section, locs_for_this_section))\n",
    "        available_y_locs = available_y_locs[len(section):]\n",
    "        for n in locs_for_this_section_dict.keys():\n",
    "            clone_y_location.loc[clone_y_location.clone == n, 'y_loc'] = locs_for_this_section_dict[n]\n",
    "    \n",
    "    tree_graph_df = pd.merge(tree_with_levels, clone_y_location)\n",
    "    \n",
    "    total_plot_height = max(1000, 75 * number_of_clones)\n",
    "    clone_prop_title = 'Clone proportions in regions'\n",
    "\n",
    "    s = [[{\"type\": \"xy\", \"rowspan\": number_of_clones}, {\"type\": \"xy\", \"rowspan\": number_of_clones}, {\"type\": \"xy\"}]]\n",
    "    for c in range(number_of_clones - 1):\n",
    "        s.append([None, None, {\"type\": \"xy\"}])\n",
    "    \n",
    "    fig = make_subplots(\n",
    "        rows=number_of_clones, cols=3,\n",
    "        column_widths=[0.1, 0.8, 0.1],\n",
    "        specs=s, horizontal_spacing=0.02, vertical_spacing=0.01,\n",
    "        subplot_titles=('', '', clone_prop_title))\n",
    "    \n",
    "    for clone_pos in tree_graph_df.y_loc:\n",
    "        hline = go.Scatter(showlegend=False,\n",
    "                           x=[-0.3, max_levels],\n",
    "                           y=[clone_pos + 0.5, clone_pos + 0.5],\n",
    "                           mode='lines',\n",
    "                           line=dict(color='Green', dash='dot'))\n",
    "        \n",
    "        fig.append_trace(hline, row=1, col=1)\n",
    "    \n",
    "    # *** plot tree ***\n",
    "    for branch in tree:\n",
    "        branch_df = tree_graph_df[tree_graph_df.clone.isin(branch)]\n",
    "        fig.append_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            name='tree',\n",
    "            x=branch_df['level'],\n",
    "            y=branch_df['y_loc'],\n",
    "            mode='lines+markers',\n",
    "            marker=dict(\n",
    "                symbol='circle',\n",
    "                color='purple',\n",
    "                size=10,\n",
    "                line=dict(\n",
    "                    color='purple',\n",
    "                    width=2)),\n",
    "            text=branch_df['clone']),\n",
    "            row=1, col=1\n",
    "        )\n",
    "    if gd_clones != []:\n",
    "        gd_df = tree_graph_df[tree_graph_df.clone.isin(gd_clones)]\n",
    "        fig.append_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            name='gd clone',\n",
    "            x=gd_df['level'],\n",
    "            y=gd_df['y_loc'],\n",
    "            mode='markers+text',\n",
    "            marker=dict(\n",
    "                symbol='circle-open',\n",
    "                color='rgb(255, 37, 62)',\n",
    "                size=16,\n",
    "                line=dict(\n",
    "                    color='rgb(255, 37, 62)',\n",
    "                    width=3)),\n",
    "            text=['WGD'] * len(gd_clones),\n",
    "            textposition=\"top right\"\n",
    "        ),\n",
    "            row=1, col=1\n",
    "        )\n",
    "    if seeding_clones != []:\n",
    "        seeding_clones_df = tree_graph_df[tree_graph_df.clone.isin(seeding_clones)]\n",
    "        fig.append_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            name='seeding clone',\n",
    "            x=seeding_clones_df['level'],\n",
    "            y=seeding_clones_df['y_loc'],\n",
    "            mode='markers+text',\n",
    "            marker=dict(\n",
    "                symbol='circle-open',\n",
    "                color='steelblue',\n",
    "                size=19,\n",
    "                line=dict(\n",
    "                    color='steelblue',\n",
    "                    width=3)),\n",
    "            text=['Seeding'],\n",
    "            textposition=\"bottom right\"\n",
    "        ),\n",
    "            row=1, col=1\n",
    "        )\n",
    "    fig.update_yaxes(\n",
    "        showgrid=True,\n",
    "        tickmode='array',\n",
    "        tickvals=list(tree_graph_df.sort_values('y_loc').y_loc),\n",
    "        ticktext=list(tree_graph_df.sort_values('y_loc').clone),\n",
    "        range=[-0.5, number_of_clones - 0.5],\n",
    "        showticklabels=True, zeroline=False, row=1, col=1)\n",
    "    fig.update_xaxes(\n",
    "        showgrid=True, zeroline=True, row=1, col=1\n",
    "    )\n",
    "    fig.update_yaxes(\n",
    "        showgrid=True,\n",
    "        tickmode='array',\n",
    "        tickvals=list(tree_graph_df.sort_values('y_loc').y_loc),\n",
    "        ticktext=list(tree_graph_df.sort_values('y_loc').clone),\n",
    "        range=[-0.5, number_of_clones - 0.5],\n",
    "        showticklabels=True, zeroline=False, row=1, col=2)\n",
    "    fig.update_xaxes(\n",
    "        showgrid=True, zeroline=True, row=1, col=2\n",
    "    )\n",
    "    \n",
    "    # *** plot clones ***\n",
    "    df = alpaca_output\n",
    "    y_limit = df['pred_CN_A'].max()\n",
    "    shapes = []\n",
    "    chr_len = get_chr_table(chr_table_file)\n",
    "    fig = plot_heat_map(alpaca_output.copy(), allele, fig, tree_graph_df, color_map,chr_table_file,mut_table)\n",
    "    for clone in df.clone.unique():\n",
    "        i = int(tree_graph_df[tree_graph_df.clone == clone]['y_loc'].iloc[0])\n",
    "        i = list(reversed(range(number_of_clones)))[i]\n",
    "        \n",
    "        fig.update_xaxes(\n",
    "            showgrid=False, row=i + 1, col=2\n",
    "        )\n",
    "        # add proportions in regions:\n",
    "        clone_cp = cp_table.loc[[clone]]\n",
    "        \n",
    "        showscale = i == len(df.clone.unique()) - 1\n",
    "        clone_proportion_heatmap = go.Heatmap(\n",
    "            z=clone_cp.values,\n",
    "            x=clone_cp.columns,\n",
    "            y=clone_cp.index,\n",
    "            text=np.round(clone_cp.values,2),\n",
    "            texttemplate=\"%{text}\",\n",
    "            textfont={\"size\":12},\n",
    "            colorscale='Blues',\n",
    "            showscale=False,\n",
    "            colorbar=dict(\n",
    "                tickfont=dict(size=12),\n",
    "                orientation='h',\n",
    "                x=0.89,\n",
    "                y=-0.1,\n",
    "                len=0.1,\n",
    "                thickness=20),\n",
    "            hoverinfo='z',\n",
    "            zauto=False,\n",
    "            zmin=0,\n",
    "            zmax=1,\n",
    "        )\n",
    "        fig.add_trace(clone_proportion_heatmap, row=i + 1, col=3)\n",
    "        \n",
    "        \n",
    "        # cleanup axes:\n",
    "        \n",
    "        fig.update_yaxes(showticklabels=False, row=i + 1, col=3)\n",
    "        fig.update_yaxes(\n",
    "            showticklabels=False, row=i + 1, col=2)\n",
    "        sample_names = clone_cp.columns\n",
    "        if i != len(df.clone.unique()) - 1:\n",
    "            fig.update_xaxes(showticklabels=False, row=i + 1, col=1)\n",
    "            \n",
    "            fig.update_xaxes(showticklabels=False, row=i + 1, col=3)\n",
    "        else:\n",
    "            sample_names = clone_cp.columns\n",
    "            # if sample names are in the long format, with tumour_id, split them:\n",
    "            if tumour_id in sample_names[0]:\n",
    "                sample_names = [x.split(f'{tumour_id}_')[1] for x in sample_names]\n",
    "            fig.update_xaxes(\n",
    "                tickmode='array',\n",
    "                ticktext=sample_names\n",
    "                ,showticklabels=True, row=i + 1, col=3)\n",
    "        \n",
    "        if i == 0:\n",
    "            fig.update_xaxes(\n",
    "                tickmode='array',\n",
    "                tickvals=chr_len['cumsum'] - (chr_len['len'] / 2),\n",
    "                ticktext=[str(x) for x in list(range(1, 23))]\n",
    "                , showticklabels=True, row=i + 1, col=2)\n",
    "    # subtitle font size:\n",
    "    fig.update_annotations(font_size=12)\n",
    "    fig.update_layout(\n",
    "        title=f'{tumour_id}<br>Allele: {allele}',\n",
    "        plot_bgcolor='rgba(255,255,255,0)',\n",
    "        autosize=False,\n",
    "        width=1600,\n",
    "        height=total_plot_height,\n",
    "        legend_tracegroupgap=10,\n",
    "        legend=dict(\n",
    "            orientation='h',\n",
    "            yanchor=\"top\",\n",
    "            y=1.4,\n",
    "            xanchor=\"left\",\n",
    "            x=0.2\n",
    "        ))\n",
    "    \n",
    "    # build legend:\n",
    "    \n",
    "    legend = {\n",
    "        '0': f'rgb{(0, 0, 255)}',\n",
    "        '1': f'rgb{(168, 168, 168)}',\n",
    "        '2': f'rgb{tuple([c * 255 for c in color_map[0]])}',\n",
    "        '3': f'rgb{tuple([c * 255 for c in color_map[1]])}',\n",
    "        '4': f'rgb{tuple([c * 255 for c in color_map[2]])}',\n",
    "        '5': f'rgb{tuple([c * 255 for c in color_map[3]])}',\n",
    "        '6': f'rgb{tuple([c * 255 for c in color_map[4]])}',\n",
    "        '7': f'rgb{tuple([c * 255 for c in color_map[5]])}',\n",
    "        '8+': f'rgb{tuple([c * 255 for c in color_map[-1]])}'\n",
    "    }\n",
    "    l_group = ['0', '0', '0',\n",
    "               '1', '1', '1',\n",
    "               '2', '2', '2', ]\n",
    "    for j, c in enumerate(legend.keys()):\n",
    "        fig.add_trace(go.Scatter(\n",
    "            legendgroup=l_group[j],\n",
    "            x=[None],\n",
    "            y=[None],\n",
    "            mode='markers',\n",
    "            name=f'{c}',\n",
    "            marker=dict(\n",
    "                color=legend[c],\n",
    "                size=10, line=dict(\n",
    "                    color='black',\n",
    "                    width=1\n",
    "                )),\n",
    "            showlegend=True\n",
    "        ), row=1, col=1)\n",
    "    \n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='3',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers\",\n",
    "        name=\"Whole Genome Doubling\",\n",
    "        marker=dict(size=10, color='rgb(255, 37, 62)', symbol='circle-open', line=dict(color='rgb(255, 37, 62)', width=3)),\n",
    "    ), row=1, col=1)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='3',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers\",\n",
    "        line=dict(\n",
    "            color='steelblue',\n",
    "            width=3),\n",
    "        name=\"Seeding Clone\",\n",
    "        marker=dict(size=10, color='steelblue', symbol='circle-open', line=dict(color='steelblue', width=3)),\n",
    "    ), row=1, col=1)\n",
    "  \n",
    "    return fig\n",
    "\n",
    "\n",
    "def plot_cpn_per_clone(tree, alpaca_output,cp_table,clone_info,wgd_calls,chr_table_file,mut_table,max_cpn_cap=8):\n",
    "    tumour_id = alpaca_output.tumour_id.iloc[0]\n",
    "    \n",
    "    # get seeding clones\n",
    "    if clone_info is not '':\n",
    "        seeding_clones = [f'clone{c}' for c in clone_info[(clone_info['SampleID']==tumour_id) & (clone_info['seedingClones'])].PyCloneCluster_SC]\n",
    "    else:\n",
    "        seeding_clones = []\n",
    "        \n",
    "    mrca = tree[0][0]\n",
    "    if wgd_calls is not None:\n",
    "        gd_clones = list(wgd_calls[(wgd_calls['tumour_id']==tumour_id) & (wgd_calls.GD==1)].clones)\n",
    "    else:\n",
    "        gd_clones = []\n",
    "    \n",
    "    # drop diploid:\n",
    "    alpaca_output = alpaca_output[alpaca_output.clone != 'diploid']\n",
    "    # add coords:\n",
    "    alpaca_output['chr'] = 'chr' + alpaca_output.segment.str.split('_', expand=True)[0]\n",
    "    alpaca_output['Start'] = alpaca_output.segment.str.split('_', expand=True)[1].astype(int)\n",
    "    alpaca_output['End'] = alpaca_output.segment.str.split('_', expand=True)[2].astype(int)\n",
    "    # modify segment positions to absolute:\n",
    "    alpaca_output = alpaca_output.merge(get_chr_table(chr_table_file))\n",
    "    alpaca_output['abs_start'] = alpaca_output['Start'] + alpaca_output['shift']\n",
    "    alpaca_output['abs_end'] = alpaca_output['End'] + alpaca_output['shift']\n",
    "    alpaca_output = alpaca_output.sort_values(['abs_start'], ascending=False)\n",
    "    alpaca_output.loc[alpaca_output['pred_CN_A'] > max_cpn_cap, 'pred_CN_A'] = max_cpn_cap\n",
    "    alpaca_output.loc[alpaca_output['pred_CN_B'] > max_cpn_cap, 'pred_CN_B'] = max_cpn_cap\n",
    "    \n",
    "    number_of_clones = len(alpaca_output.clone.unique())\n",
    "    max_levels = max([len(b) for b in tree])\n",
    "    tree_with_levels = [dict(zip(b, range(0, len(b)))) for b in tree]\n",
    "    tree_with_levels = pd.concat([pd.DataFrame(tree_with_levels[x], index=[0]).transpose() for x, _ in enumerate(tree_with_levels)]).reset_index().drop_duplicates().rename(\n",
    "        columns={'index': 'clone', 0: 'level'})\n",
    "    \n",
    "    # make empty y_loc df\n",
    "    clone_y_location = dict(zip(alpaca_output.clone.unique(), range(0, number_of_clones)))\n",
    "    clone_y_location = pd.DataFrame(clone_y_location, index=[0]).transpose().reset_index().rename(columns={'index': 'clone', 0: 'y_loc'})\n",
    "    clone_y_location['y_loc'] = 100\n",
    "    # find sections:\n",
    "    # section is a part of a path that requires its own horizontal space on the final graph\n",
    "    ori_tree = tree.copy()\n",
    "    sections = [[tree[0][0]]]\n",
    "    while len(flat_list(ori_tree)) > 1:\n",
    "        for i, branch in enumerate(ori_tree):\n",
    "            if (branch != []) and (set(flat_list(sections)) != set(flat_list(tree))):\n",
    "                branching_clones = list(pd.Series(flat_list(ori_tree)).value_counts()[pd.Series(flat_list(ori_tree)).value_counts() > 1].index)\n",
    "                if branching_clones == []:\n",
    "                    branching_clones = [mrca]\n",
    "                section_start = max([branch.index(x) for x in branching_clones if x in branch]) + 1\n",
    "                section = branch[section_start:]\n",
    "                ori_tree[ori_tree.index(branch)] = branch[:section_start]\n",
    "                ori_tree = get_unique_lists(ori_tree)\n",
    "                if section != []:\n",
    "                    sections.append(section)\n",
    "    # order sections according to proximity\n",
    "    # start with on of the longest paths\n",
    "    section_termini = [x[-1] for x in sections]\n",
    "    initial_node = [s for s in sections if len(s) == max([len(x) for x in sections])][0][-1]\n",
    "    # simplify tree graph:\n",
    "    simple_tree = [[clone for clone in branch if clone in section_termini] for branch in tree]\n",
    "    \n",
    "    edges = get_tree_edges(simple_tree)\n",
    "    \n",
    "    G = nx.Graph()\n",
    "    for edge in edges:\n",
    "        G.add_edge(edge[0], edge[1], weight=0)\n",
    "    \n",
    "    nodes = [initial_node] + [x for x in list(G.nodes) if x is not initial_node]\n",
    "    distance_to_nodes = dict(nx.all_pairs_shortest_path_length(G))\n",
    "    processed_nodes = [[s for s in sections if s[-1] == initial_node][0][-1]]\n",
    "    \n",
    "    while len(nodes) > 0:\n",
    "        node = processed_nodes[-1]\n",
    "        neighbours = pd.DataFrame(distance_to_nodes[node], index=['val']).transpose()\n",
    "        neighbours.drop(inplace=True, index=processed_nodes)\n",
    "        try:\n",
    "            neighbours.drop(inplace=True, index=node)\n",
    "        except KeyError:\n",
    "            pass\n",
    "        neighbours = neighbours.reset_index().rename(columns={'index': 'clone'}).merge(tree_with_levels, how='left', on='clone')\n",
    "        neighbours['level'] = neighbours['level'].astype(int)\n",
    "        if len(neighbours) > 0:\n",
    "            # if there are to neighbours equaly close, choose the one which has higher level (i.e. is deeper in the tree)\n",
    "            # to do that, multiply proximity by negaive level:\n",
    "            closest_candidates = neighbours[neighbours.val == min(neighbours.val)]\n",
    "            closest_neighbours = closest_candidates[closest_candidates.level == max(closest_candidates.level)]\n",
    "            closest_neighbour = closest_neighbours.clone.values[0]\n",
    "            if closest_neighbour not in processed_nodes:\n",
    "                processed_nodes.append(closest_neighbour)\n",
    "        nodes = [n for n in nodes if n != node]\n",
    "    \n",
    "    sorted_sections_raw = []\n",
    "    for node in processed_nodes:\n",
    "        sorted_sections_raw.append([s for s in sections if s[-1] == node][0])\n",
    "    \n",
    "    # join single-clone sections below MRCA with their descendants:\n",
    "    sorted_sections = []\n",
    "    below_MRCA = True\n",
    "    skip_element = False\n",
    "    for ss in sorted_sections_raw:\n",
    "        if skip_element:\n",
    "            skip_element = False\n",
    "            continue\n",
    "        if mrca in ss:\n",
    "            below_MRCA = False\n",
    "        if below_MRCA:\n",
    "            if len(ss) == 1:\n",
    "                joined = ss + sorted_sections_raw[sorted_sections_raw.index(ss) + 1]\n",
    "                sorted_sections.append(joined)\n",
    "                skip_element = True\n",
    "            else:\n",
    "                sorted_sections.append(ss)\n",
    "        else:\n",
    "            sorted_sections.append(ss)\n",
    "    \n",
    "    # assign y location on the plot to sorted sections:\n",
    "    available_y_locs = range(0, number_of_clones)\n",
    "    below_MRCA = True\n",
    "    for section in sorted_sections:\n",
    "        if mrca in section:\n",
    "            below_MRCA = False\n",
    "        locs_for_this_section = list(available_y_locs[:len(section)])\n",
    "        if below_MRCA:\n",
    "            locs_for_this_section = list(reversed(locs_for_this_section))\n",
    "        locs_for_this_section_dict = dict(zip(section, locs_for_this_section))\n",
    "        available_y_locs = available_y_locs[len(section):]\n",
    "        for n in locs_for_this_section_dict.keys():\n",
    "            clone_y_location.loc[clone_y_location.clone == n, 'y_loc'] = locs_for_this_section_dict[n]\n",
    "    \n",
    "    tree_graph_df = pd.merge(tree_with_levels, clone_y_location)\n",
    "    \n",
    "    clone_prop_title = 'Clone proportions in regions'\n",
    "\n",
    "    s = [[{\"type\": \"xy\", \"rowspan\": number_of_clones}, {\"type\": \"xy\"}, {\"type\": \"xy\"}]]\n",
    "    for c in range(number_of_clones - 1):\n",
    "        s.append([None, {\"type\": \"xy\"}, {\"type\": \"xy\"}])\n",
    "    \n",
    "    fig = make_subplots(\n",
    "        rows=number_of_clones, cols=3,\n",
    "        column_widths=[0.1, 0.8, 0.1],\n",
    "        specs=s, horizontal_spacing=0.02, vertical_spacing=0.01,\n",
    "        subplot_titles=('', '', clone_prop_title))\n",
    "    \n",
    "    for clone_pos in tree_graph_df.y_loc:\n",
    "        hline = go.Scatter(\n",
    "            showlegend=False,\n",
    "            x=[-0.3, max_levels],\n",
    "            y=[clone_pos + 0.5, clone_pos + 0.5],\n",
    "            mode='lines',\n",
    "            line=dict(color='Green', dash='dot'))\n",
    "        \n",
    "        fig.append_trace(hline, row=1, col=1)\n",
    "    \n",
    "    # *** plot tree ***\n",
    "    for branch in tree:\n",
    "        branch_df = tree_graph_df[tree_graph_df.clone.isin(branch)]\n",
    "        fig.append_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            name='tree',\n",
    "            x=branch_df['level'],\n",
    "            y=branch_df['y_loc'],\n",
    "            mode='lines+markers',\n",
    "            marker=dict(\n",
    "                symbol='circle',\n",
    "                color='purple',\n",
    "                size=10,\n",
    "                line=dict(\n",
    "                    color='purple',\n",
    "                    width=2)),\n",
    "            text=branch_df['clone']),\n",
    "            row=1, col=1\n",
    "        )\n",
    "    if gd_clones != []:\n",
    "        gd_df = tree_graph_df[tree_graph_df.clone.isin(gd_clones)]\n",
    "        fig.append_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            name='gd clone',\n",
    "            x=gd_df['level'],\n",
    "            y=gd_df['y_loc'],\n",
    "            mode='markers+text',\n",
    "            marker=dict(\n",
    "                symbol='circle-open',\n",
    "                color='rgb(255, 37, 62)',\n",
    "                size=16,\n",
    "                line=dict(\n",
    "                    color='rgb(255, 37, 62)',\n",
    "                    width=3)),\n",
    "            text=['WGD'] * len(gd_clones),\n",
    "            textposition=\"top right\"\n",
    "        ),\n",
    "            row=1, col=1\n",
    "        )\n",
    "    if seeding_clones != []:\n",
    "        seeding_clones_df = tree_graph_df[tree_graph_df.clone.isin(seeding_clones)]\n",
    "        fig.append_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            name='seeding clone',\n",
    "            x=seeding_clones_df['level'],\n",
    "            y=seeding_clones_df['y_loc'],\n",
    "            mode='markers+text',\n",
    "            marker=dict(\n",
    "                symbol='circle-open',\n",
    "                color='steelblue',\n",
    "                size=19,\n",
    "                line=dict(\n",
    "                    color='steelblue',\n",
    "                    width=3)),\n",
    "            text=['Seeding'],\n",
    "            textposition=\"bottom right\"\n",
    "        ),\n",
    "            row=1, col=1\n",
    "        )\n",
    "    fig.update_yaxes(\n",
    "        showgrid=False,\n",
    "        tickmode='array',\n",
    "        tickvals=list(tree_graph_df.sort_values('y_loc').y_loc),\n",
    "        ticktext=list(tree_graph_df.sort_values('y_loc').clone),\n",
    "        range=[-0.3, number_of_clones - 0.7],\n",
    "        showticklabels=True, zeroline=False, row=1, col=1)\n",
    "    fig.update_xaxes(\n",
    "        showgrid=False, zeroline=False, row=1, col=1\n",
    "    )\n",
    "    total_plot_height = max(1000, 75 * number_of_clones)\n",
    "    # *** plot clones ***\n",
    "    df = alpaca_output\n",
    "    shapes = []\n",
    "    chr_len = get_chr_table(chr_table_file)\n",
    "    driver_mutations = get_driver_mutations(mut_table,tumour_id,chr_table_file)\n",
    "    for clone in df.clone.unique():\n",
    "        i = int(tree_graph_df[tree_graph_df.clone == clone]['y_loc'].iloc[0])\n",
    "        i = list(reversed(range(number_of_clones)))[i]\n",
    "        y_limit = df['pred_CN_A'].max()\n",
    "        clone_df = df[df.clone == clone]\n",
    "        clone_df = clone_df[['abs_start', 'abs_end', 'pred_CN_A', 'pred_CN_B']].drop_duplicates()\n",
    "        clone_df = clone_df.sort_values('abs_start', ascending=True)\n",
    "        clone_df['space'] = None\n",
    "        x = flat_list([[v[1]['abs_start'], v[1]['abs_end'], v[1]['space']] for v in clone_df.iterrows()])\n",
    "        ya = flat_list([[v[1]['pred_CN_A'], v[1]['pred_CN_A'], v[1]['space']] for v in clone_df.iterrows()])\n",
    "        yb = flat_list([[v[1]['pred_CN_B'], v[1]['pred_CN_B'], v[1]['space']] for v in clone_df.iterrows()])\n",
    "        clone_df_sameAB = clone_df[clone_df['pred_CN_A'] == clone_df['pred_CN_B']]\n",
    "        yab = flat_list([[v[1]['pred_CN_B'], v[1]['pred_CN_B'], v[1]['space']] for v in clone_df_sameAB.iterrows()])\n",
    "        clone_mutations = driver_mutations[driver_mutations.clone==clone]\n",
    "        \n",
    "        fig.add_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            x=x,\n",
    "            y=ya,\n",
    "            mode='lines',\n",
    "            line=dict(\n",
    "                color='rgb(255, 164, 0)',\n",
    "                width=1)),\n",
    "            row=i + 1, col=2\n",
    "        )\n",
    "        \n",
    "        fig.add_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            x=x,\n",
    "            y=yb,\n",
    "            mode='lines',\n",
    "            line=dict(\n",
    "                color='rgb(0, 128, 128)',\n",
    "                width=1)),\n",
    "            row=i + 1, col=2\n",
    "        )\n",
    "        \n",
    "        fig.add_trace(go.Scatter(\n",
    "            showlegend=False,\n",
    "            x=x,\n",
    "            y=yab,\n",
    "            mode='lines',\n",
    "            line=dict(\n",
    "                color='rgb(187, 206, 186)',\n",
    "                width=1)),\n",
    "            row=i + 1, col=2\n",
    "        )\n",
    "        \n",
    "        # add loss/gain areas:\n",
    "        ancestor_clone = find_parent(clone, tree)\n",
    "        if ancestor_clone == 'diploid':\n",
    "            ancestor_df = clone_df.copy()\n",
    "            ancestor_df.loc[:, 'pred_CN_A'] = 1\n",
    "            ancestor_df.loc[:, 'pred_CN_B'] = 1\n",
    "        else:\n",
    "            ancestor_df = df[df.clone == ancestor_clone]\n",
    "            ancestor_df = ancestor_df[['abs_start', 'abs_end', 'pred_CN_A', 'pred_CN_B']].drop_duplicates()\n",
    "            ancestor_df = ancestor_df.sort_values('abs_start', ascending=True)\n",
    "            ancestor_df['space'] = None\n",
    "        gain_loss_df = pd.merge(clone_df, ancestor_df, on=['abs_start', 'abs_end', 'space'])\n",
    "        gain_loss_df['gain'] = (gain_loss_df['pred_CN_A_x'] > gain_loss_df['pred_CN_A_y']) | (gain_loss_df['pred_CN_B_x'] > gain_loss_df['pred_CN_B_y'])\n",
    "        gain_loss_df['loss'] = (gain_loss_df['pred_CN_A_x'] < gain_loss_df['pred_CN_A_y']) | (gain_loss_df['pred_CN_B_x'] < gain_loss_df['pred_CN_B_y'])\n",
    "        \n",
    "        gain_and_loss_df = gain_loss_df[(gain_loss_df['loss'] == True) & (gain_loss_df['gain'] == True)]\n",
    "        gain_df = gain_loss_df[(gain_loss_df['loss'] == False) & (gain_loss_df['gain'] == True)]\n",
    "        loss_df = gain_loss_df[(gain_loss_df['loss'] == True) & (gain_loss_df['gain'] == False)]\n",
    "        D = {'gain_and_loss_df': gain_and_loss_df, 'gain_df': gain_df, 'loss_df': loss_df}\n",
    "        \n",
    "        xxyy = {}\n",
    "        \n",
    "        for d in D.items():\n",
    "            if len(d[1] > 0):\n",
    "                xxyy[f'x_{d[0]}'] = flat_list([[v[1]['abs_start'], v[1]['abs_start'], v[1]['abs_end'], v[1]['abs_end'], v[1]['abs_start'], v[1]['space']] for v in d[1].iterrows()])\n",
    "                xxyy[f'y_{d[0]}'] = flat_list([[0, y_limit, y_limit, 0, 0, v[1]['space']] for v in d[1].iterrows()])\n",
    "        \n",
    "        try:\n",
    "            fig.add_trace(go.Scatter(\n",
    "                showlegend=False,\n",
    "                x=xxyy['x_gain_df'],\n",
    "                y=xxyy['y_gain_df'],\n",
    "                mode='lines',\n",
    "                fill=\"toself\",\n",
    "                line=dict(\n",
    "                    color='rgb(255, 230, 229,0.5)',\n",
    "                    width=1)),\n",
    "                row=i + 1, col=2\n",
    "            )\n",
    "        except KeyError:\n",
    "            pass\n",
    "        try:\n",
    "            fig.add_trace(go.Scatter(\n",
    "                showlegend=False,\n",
    "                x=xxyy['x_loss_df'],\n",
    "                y=xxyy['y_loss_df'],\n",
    "                mode='lines',\n",
    "                fill=\"toself\",\n",
    "                line=dict(\n",
    "                    color='rgb(230, 229, 255,0.5)',\n",
    "                    width=1)),\n",
    "                row=i + 1, col=2\n",
    "            )\n",
    "        except KeyError:\n",
    "            pass\n",
    "        try:\n",
    "            fig.add_trace(go.Scatter(\n",
    "                showlegend=False,\n",
    "                x=xxyy['x_gain_and_loss_df'],\n",
    "                y=xxyy['y_gain_and_loss_df'],\n",
    "                mode='lines',\n",
    "                fill=\"toself\",\n",
    "                line=dict(\n",
    "                    color='rgb(230, 207, 232,0.5)',\n",
    "                    width=1)),\n",
    "                row=i + 1, col=2\n",
    "            )\n",
    "        except KeyError:\n",
    "            pass\n",
    "        fig.add_trace(go.Scatter(\n",
    "            x=clone_mutations.abs_position,\n",
    "            y=[0]*len(clone_mutations),\n",
    "            hovertext=clone_mutations.gene,\n",
    "            mode='markers',showlegend=False,marker=dict(color='teal')), row=i + 1, col=2)\n",
    "        #add static annotations:\n",
    "        '''\n",
    "        for index, row in clone_mutations.iterrows():\n",
    "            fig.add_annotation(\n",
    "                x=row['abs_position'],\n",
    "                y=[0]*len(clone_mutations),\n",
    "                text=row['gene'],\n",
    "                font=dict(size=20),\n",
    "                showarrow=False,\n",
    "                yshift=0,\n",
    "                xref=f\"x{i+1}\",\n",
    "                yref=f\"y2\"\n",
    "            )\n",
    "         '''   \n",
    "        fig.update_xaxes(\n",
    "            showgrid=False, row=i + 1, col=2\n",
    "        )\n",
    "        \n",
    "        # add proportions in regions:\n",
    "        clone_cp = cp_table.loc[[clone]]\n",
    "        \n",
    "        showscale = i == len(df.clone.unique()) - 1\n",
    "        clone_proportion_heatmap = go.Heatmap(\n",
    "            z=clone_cp.values,\n",
    "            x=clone_cp.columns,\n",
    "            y=clone_cp.index,\n",
    "            text=np.round(clone_cp.values,2),\n",
    "            texttemplate=\"%{text}\",\n",
    "            textfont={\"size\":12},\n",
    "            colorscale='Blues',\n",
    "            showscale=False,\n",
    "            colorbar=dict(\n",
    "                tickfont=dict(size=12),\n",
    "                orientation='h',\n",
    "                x=0.89,\n",
    "                y=-0.1,\n",
    "                len=0.1,\n",
    "                thickness=20),\n",
    "            hoverinfo='z',\n",
    "            zauto=False,\n",
    "            zmin=0,\n",
    "            zmax=1,\n",
    "        )\n",
    "        fig.add_trace(clone_proportion_heatmap, row=i + 1, col=3)\n",
    "        \n",
    "        # cleanup axes:\n",
    "        for chromosome_line in chr_len['cumsum']:\n",
    "            fig.add_trace(go.Scatter(\n",
    "                x=[chromosome_line, chromosome_line],\n",
    "                y=[0, y_limit],\n",
    "                mode='lines', line=dict(color='black', width=1, dash='dot'), showlegend=False), row=i + 1, col=2)\n",
    "        fig.update_yaxes(showticklabels=False, row=i + 1, col=3)\n",
    "        #fig.update_yaxes(showticklabels=False, row=i + 1, col=2)\n",
    "        sample_names = clone_cp.columns\n",
    "        if i != len(df.clone.unique()) - 1:\n",
    "            fig.update_xaxes(showticklabels=False, row=i + 1, col=1)\n",
    "            fig.update_xaxes(showticklabels=False, row=i + 1, col=2)\n",
    "            fig.update_xaxes(showticklabels=False, row=i + 1, col=3)\n",
    "        else:\n",
    "            fig.update_xaxes(\n",
    "                tickmode='array',\n",
    "                tickvals=chr_len['cumsum'] - (chr_len['len'] / 2),\n",
    "                ticktext=[str(x) for x in list(range(1, 23))]\n",
    "                , showticklabels=True, row=i + 1, col=2)\n",
    "            sample_names = clone_cp.columns\n",
    "            # if sample names are in the long format, with tumour_id, split them:\n",
    "            if tumour_id in sample_names[0]:\n",
    "                sample_names = [x.split(f'{tumour_id}_')[1] for x in sample_names]\n",
    "            fig.update_xaxes(\n",
    "                tickmode='array',\n",
    "                ticktext=sample_names\n",
    "                ,showticklabels=True, row=i + 1, col=3)\n",
    "        \n",
    "\n",
    "            \n",
    "    \n",
    "    # subtitle font size:\n",
    "    fig.update_annotations(font_size=12)\n",
    "    fig.update_layout(\n",
    "        title=f'{tumour_id}',\n",
    "        plot_bgcolor='rgb(255,255,255)',\n",
    "        autosize=False,\n",
    "        width=1600,\n",
    "        height=total_plot_height,\n",
    "        legend_tracegroupgap=10,\n",
    "        legend=dict(\n",
    "            orientation='h',\n",
    "            yanchor=\"top\",\n",
    "            y=1.4,\n",
    "            xanchor=\"left\",\n",
    "            x=0.2\n",
    "        ))\n",
    "    \n",
    "    # build legend:\n",
    "    # gain loss areas:\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='1',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers\",\n",
    "        name=\"gain relative to parent\",\n",
    "        marker=dict(size=10, color='rgb(255, 230, 229)', symbol='square'),\n",
    "    ), row=1, col=1)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='1',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers\",\n",
    "        name=\"loss relative to parent\",\n",
    "        marker=dict(size=10, color='rgb(230, 229, 255)', symbol='square'),\n",
    "    ), row=1, col=1)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='1',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers\",\n",
    "        name=\"gain and loss relative to parent\",\n",
    "        marker=dict(size=10, color='rgb(230, 207, 232)', symbol='square'),\n",
    "    ), row=1, col=1)\n",
    "    # cpn lines:\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='2',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers+lines\",\n",
    "        name=\"A allele\",\n",
    "        marker=dict(size=10, color='rgb(255, 164, 0)', symbol='line-ew'),\n",
    "    ), row=1, col=1)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='2',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers+lines\",\n",
    "        name=\"B allele\",\n",
    "        marker=dict(size=10, color='rgb(0, 128, 128)', symbol='line-ew'),\n",
    "    ), row=1, col=1)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='2',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers+lines\",\n",
    "        name=\"A and B allele\",\n",
    "        marker=dict(size=10, color='rgb(187, 206, 186)', symbol='line-ew'),\n",
    "    ), row=1, col=1)\n",
    "    \n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='3',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers\",\n",
    "        name=\"Whole Genome Doubling\",\n",
    "        marker=dict(size=10, color='rgb(255, 37, 62)', symbol='circle-open', line=dict(color='rgb(255, 37, 62)', width=3)),\n",
    "    ), row=1, col=1)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        legendgroup='3',\n",
    "        x=[None],\n",
    "        y=[None],\n",
    "        mode=\"markers\",\n",
    "        line=dict(\n",
    "            color='steelblue',\n",
    "            width=3),\n",
    "        name=\"Seeding Clone\",\n",
    "        marker=dict(size=10, color='steelblue', symbol='circle-open', line=dict(color='steelblue', width=3)),\n",
    "    ), row=1, col=1)\n",
    "    return fig\n",
    "\n",
    "\n",
    "def plot_heat_map(patient_output, allele, fig, tree_graph_df, color_map,chr_table_file,mut_table):\n",
    "    clones = tree_graph_df.sort_values('y_loc', ascending=True).clone\n",
    "    clone_number = len(clones)\n",
    "    chromosome_table = get_chr_table(chr_table_file)\n",
    "    tumour_id = patient_output.tumour_id.iloc[0]\n",
    "    patient_output = clean_output(patient_output)\n",
    "    patient_output['predicted_cpn'] = patient_output[f'pred_CN_{allele}']\n",
    "    # patient_output['fractional_cpn'] = patient_output[f'mphase{allele}_allele']\n",
    "    patient_output = patient_output.merge(tree_graph_df)\n",
    "    patient_output = patient_output.sort_values('y_loc')\n",
    "    color_map = sns.color_palette(\"rocket\", 7, as_cmap=False)\n",
    "    color_map.reverse()\n",
    "    \n",
    "    def getColour(cp_state):\n",
    "        if cp_state >= 8:\n",
    "            return f'rgb{tuple([c * 255 for c in color_map[-1]])}'\n",
    "        else:\n",
    "            colours = {\n",
    "                0: (0, 0, 1), 1: (0.66, 0.66, 0.66), 2: color_map[0], 3: color_map[1], 4: color_map[2], 5: color_map[3], 6: color_map[4],\n",
    "                7: color_map[5]\n",
    "            }\n",
    "            \n",
    "            return f'rgb{tuple([c * 255 for c in colours[cp_state]])}'\n",
    "    \n",
    "    # fig = go.Figure(layout_xaxis_range=[0, 3.2 * 10 ** 9], layout_yaxis_range=[0, len(clones) + 1])\n",
    "    driver_mutations = get_driver_mutations(mut_table,tumour_id,chr_table_file) if mut_table is not None else None\n",
    "    for clone_index, clone_name in enumerate(clones):\n",
    "        clone_df = patient_output[patient_output['clone'] == clone_name]\n",
    "        for cp_state in clone_df.predicted_cpn.unique():\n",
    "            clone_df_cp_state = clone_df[clone_df.predicted_cpn == cp_state]\n",
    "            segments_predicted = [[tuple([row[1]['abs_start'], clone_index - 0.5]), tuple([row[1]['abs_end'], clone_index + 0.5])] for row in clone_df_cp_state.iterrows()]\n",
    "            segments_predicted_unique = []\n",
    "            for x in segments_predicted:\n",
    "                if x not in segments_predicted_unique:\n",
    "                    segments_predicted_unique.append(x)\n",
    "            cpn_color = getColour(cp_state)\n",
    "            for rectangle in segments_predicted_unique:\n",
    "                segment = clone_df_cp_state[clone_df_cp_state['abs_start'] == rectangle[0][0]].segment.unique()[0]\n",
    "                fig.add_trace(go.Scatter(\n",
    "                    showlegend=False,\n",
    "                    x=[rectangle[0][0], rectangle[0][0], rectangle[1][0], rectangle[1][0]],\n",
    "                    y=[rectangle[0][1], rectangle[1][1], rectangle[1][1], rectangle[0][1]],\n",
    "                    # y=[rectangle[0][1], rectangle[0][1] + 1, rectangle[0][1] + 1, rectangle[0][1]],\n",
    "                    fill='toself',\n",
    "                    mode='lines',\n",
    "                    fillcolor=cpn_color,\n",
    "                    line_color=cpn_color,\n",
    "                    name=f'clone: {rectangle[0][1]}, seg: {segment}'), row=1, col=2)\n",
    "        if driver_mutations is not None:\n",
    "            clone_mutations = driver_mutations[driver_mutations.clone==clone_name]\n",
    "            # add mutation markers\n",
    "            fig.add_trace(go.Scatter(\n",
    "                x=clone_mutations.abs_position,\n",
    "                y=[clone_index]*len(clone_mutations),\n",
    "                mode='markers',showlegend=False,marker=dict(color='teal')), row=1, col=2)\n",
    "            #add static annotations:\n",
    "            for index, row in clone_mutations.iterrows():\n",
    "                fig.add_annotation(\n",
    "                    x=row['abs_position'],\n",
    "                    y=clone_index,\n",
    "                    text=row['gene'],\n",
    "                    font=dict(size=50),\n",
    "                    showarrow=False,\n",
    "                    yshift=10,\n",
    "                    row=1, col=2\n",
    "                )\n",
    "        \n",
    "    chromosomes = [[tuple([row[1]['cumsum'], 0]), tuple([row[1]['cumsum'], clone_number])] for row in chromosome_table.iterrows()]\n",
    "    for chromosome_line in chromosomes:\n",
    "        fig.add_trace(go.Scatter(\n",
    "            x=[chromosome_line[0][0], chromosome_line[0][0]],\n",
    "            y=[-0.5, clone_number + 0.5],\n",
    "            mode='lines', line=dict(color='black', width=1, dash='dash'), showlegend=False), row=1, col=2)\n",
    "    '''fig.update_layout(\n",
    "        xaxis=dict(\n",
    "            tickmode='array',\n",
    "            tickvals=chromosome_table['half_lengths_abs'],\n",
    "            ticktext=[str(x) for x in list(range(1, 23))]\n",
    "        ),\n",
    "        yaxis=dict(\n",
    "            tickfont=dict(size=8),\n",
    "            tickmode='array',\n",
    "            tickvals=np.arange(1.5, clone_number + 1.5, 1),\n",
    "            ticktext=[c for c in clones]\n",
    "        )\n",
    "    )\n",
    "    fig.update_yaxes(showgrid=False, gridwidth=1, gridcolor='LightPink')\n",
    "    fig.update_layout(showlegend=True, plot_bgcolor='rgba(0,0,0,0)')'''\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = '../..'\n",
    "\n",
    "patient_outputs = f'{output_directory}/output/primary/primary_default/patient_outputs'\n",
    "mut_table = f'{output_directory}/_assets/mutTable/primary_mutTable.csv'\n",
    "wgd_calls = f'{output_directory}/output/primary/primary_default/cohort_outputs/wgd_calls.csv'\n",
    "chr_table_file = f'{output_directory}/_assets/chr_len.csv'\n",
    "clone_info = ''\n",
    "\n",
    "tumour_id = 'CRUK0717'\n",
    "tree = read_tree_json(f'{patient_outputs}/{tumour_id}/tree_paths.json')\n",
    "alpaca_output = pd.read_csv(f'{patient_outputs}/{tumour_id}/combined.csv')\n",
    "cp_table = pd.read_csv(f'{patient_outputs}/{tumour_id}/cp_table.csv', index_col='clone')\n",
    "cp_table.columns = [x[2:] for x in cp_table.columns if 'CRUK' in x]\n",
    "wgd_calls = pd.read_csv(wgd_calls)\n",
    "allele='A'\n",
    "high_diversity_heatmap_A = plot_heatmap_with_tree(tree, alpaca_output,cp_table,clone_info,wgd_calls,chr_table_file,mut_table,allele=allele)\n",
    "high_diversity_heatmap_A.write_image(f'{output_directory}/figures/Fig5b_{tumour_id}_{allele}_high_diversity_heatmap.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_diversity_heatmap_A.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_outputs = f'{output_directory}/output/primary/primary_default/patient_outputs'\n",
    "mut_table = f'{output_directory}/_assets/mutTable/primary_mutTable.csv'\n",
    "wgd_calls = f'{output_directory}/output/primary/primary_default/cohort_outputs/wgd_calls.csv'\n",
    "chr_table_file = f'{output_directory}/_assets/chr_len.csv'\n",
    "clone_info = ''\n",
    "tumour_id = 'CRUK0003'\n",
    "tree = read_tree_json(f'{patient_outputs}/{tumour_id}/tree_paths.json')\n",
    "alpaca_output = pd.read_csv(f'{patient_outputs}/{tumour_id}/combined.csv')\n",
    "cp_table = pd.read_csv(f'{patient_outputs}/{tumour_id}/cp_table.csv', index_col='clone')\n",
    "cp_table.columns = [x[2:] for x in cp_table.columns if 'CRUK' in x]\n",
    "wgd_calls = pd.read_csv(wgd_calls)\n",
    "allele='A'\n",
    "low_diversity_heatmap_A = plot_heatmap_with_tree(tree, alpaca_output,cp_table,clone_info,wgd_calls,chr_table_file,mut_table,allele=allele)\n",
    "low_diversity_heatmap_A.write_image(f'{output_directory}/figures/Fig5b_{tumour_id}_{allele}_low_diversity_heatmap.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_diversity_heatmap_A.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
